### PR TITLE
[boo] refactor numeric driver towards generalization

### PR DESCRIPTION
**Pull request by ftynse**
Move the main logic of the numerics checker from `conv_exports` to `exports` and generalize it to support cases beyond convolution by querying the op signature. Keep the `numerics_driver.py` shim under `conv_exports` for now as doing that would require a parser registry.

### CODE REVIEW

File: iree/turbine/kernel/boo/exports/signature.py [hunk 1]
```diff
@@ -38,3 +56,55 @@ def get_nn_module(self, **kwargs) -> torch.nn.Module:
...
+
+    @abstractmethod
+    def make_signature_copy_for_forward(self) -> "OpSignature":
+        """Returns a copy of this signature, but set up for forward mode."""
+        ...
+
+    @abstractmethod
+    def get_arg_index_for_backward(self) -> int | None:
+        """For a backward mode, returns the index of the argument the derivative
+        of which is computed by the kernel with this signature.
```
Comment (zjgarvey) [inline iree/turbine/kernel/boo/exports/signature.py:75]: small nit: we aren't computing the derivative of an input. I'd maybe change `"... the derivative of which ..."` to `"... whose back-propagated gradient ..."`.

File: iree/turbine/kernel/boo/driver/numerics.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+        # Compute backward values. Only propagate derivatives through the "main"
+        # result of the operation. Otherwise we'd need to handle a full Jacobian
+        # matrix here with significantly more complexity.
+        main_result_ref_gpu = results_ref_gpu[sig.main_result_index]
+        main_result_ref_gpu.retain_grad()
+        # TODO: this loss function looks like it can easily explode given
+        # that values are not normalized.
+        loss = torch.sum((main_result_ref_gpu * main_result_ref_gpu) / 2)
+        loss.backward()
```
[the remark this answers is not in the document]
Comment (ftynse) [reply iree/turbine/kernel/boo/driver/numerics.py:133]: Let's have that as #1021, also updated the todo.

File: iree/turbine/kernel/boo/driver/numerics.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        }
+
+    return results
+
+
+from pathlib import Path
+
+
+# TODO: consider extracting kind from the parser class instead of passing a string
+def _load_commands(commands_file: str, *, kind: str) -> list[str]:
```
[the remark this answers is not in the document]
Comment (ftynse) [reply iree/turbine/kernel/boo/driver/numerics.py:169]: Sure, moved that function too

**Review by zjgarvey — APPROVED** [review]
